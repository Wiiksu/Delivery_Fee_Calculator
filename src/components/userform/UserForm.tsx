import { useForm, SubmitHandler } from "react-hook-form";
import { yupResolver } from "@hookform/resolvers/yup";
import { formValidatorSchema } from "../../helpers/FormValidation";
import { InputField } from "../inputfield/InputField";
import SubmitBtn from "../submitbtn/SubmitBtn";
import {
  UserFormProps,
  UserFormReturnType,
  UserFormType,
} from "../../helpers/types";
import { surchargeCalculator } from "../../helpers/SurchargeCalculator";

export default function UserForm({ saveCalculationResults }: UserFormProps) {
  const {
    handleSubmit,
    register,
    formState: { errors },
    resetField,
  } = useForm<UserFormType>({
    mode: "all",
    resolver: yupResolver(formValidatorSchema),
    reValidateMode: "onChange",
  });

  const onSubmit: SubmitHandler<UserFormType> = (data: UserFormType) => {
    try {
      const results: UserFormReturnType = surchargeCalculator(data);
      saveCalculationResults(results);
      resetField("cartValue");
      resetField("deliveryDistance");
      resetField("numberOfItems");
      resetField("orderTime");
    } catch (error) {
      console.error("Error in surcharge calculation:");
      console.error(error);
    }
  };

  return (
    <section className="bg-purple-400 px-6 pb-6 pt-2 rounded shadow-lg h-fit w-[22rem] md:w-[27rem]">
      <h2 className="text-center text-purple-50 text-2xl font-bold py-2 mt-3 mb-4 bg-purple-500 border-2 border-purple-50 border-opacity-60 rounded-md select-none">
        Order Details
      </h2>
      <form
        noValidate
        className="flex flex-col gap-1"
        onSubmit={handleSubmit(onSubmit)}
      >
        <InputField
          {...register("cartValue")}
          data-testid="cartValue"
          label="Cart Value"
          type="text"
          errorMessage={errors.cartValue?.message}
        />
        <InputField
          {...register("deliveryDistance")}
          data-testid="deliveryDistance"
          label="Delivery distance"
          type="text"
          errorMessage={errors.deliveryDistance?.message}
        />
        <InputField
          {...register("numberOfItems")}
          data-testid="numberOfItems"
          label="Amount of items"
          type="text"
          errorMessage={errors.numberOfItems?.message}
        />
        <InputField
          {...register("orderTime")}
          data-testid="orderTime"
          label="Time"
          type="datetime-local"
          errorMessage={errors.orderTime?.message}
        />
        <div className="flex flex-col mt-2 gap-3">
          <SubmitBtn title="Calculate Delivery Fee" type="submit" />
        </div>
      </form>
    </section>
  );
}
