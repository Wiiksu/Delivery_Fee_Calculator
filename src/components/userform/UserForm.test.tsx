import { expect, it, describe, vi, beforeEach } from "vitest";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import UserForm from "./UserForm";

// Testing if invalid value(s) prevent from submitting
const MOCK_INVALID_VALUES = {
  cartValue: -1,
  deliveryDistance: 1000,
  numberOfItems: 4,
  orderTime: "2024-02-23T15:00",
};

// Testing if valid values pass on after submitting the form
const MOCK_VALID_VALUES = {
  cartValue: 9,
  deliveryDistance: 1500,
  numberOfItems: 13,
  orderTime: "2024-02-23T15:00",
};

describe("<UserForm />", () => {
  const mockSaveCalculationResults = vi.fn();

  beforeEach(() => {
    render(<UserForm saveCalculationResults={mockSaveCalculationResults} />);
  });

  it("Should render correctly to the snapshot", () => {
    const { asFragment } = render(
      <UserForm saveCalculationResults={mockSaveCalculationResults} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("Renders form fields correctly", () => {
    expect(screen.getByTestId("cartValue")).toBeInTheDocument();
    expect(screen.getByTestId("deliveryDistance")).toBeInTheDocument();
    expect(screen.getByTestId("numberOfItems")).toBeInTheDocument();
    expect(screen.getByTestId("orderTime")).toBeInTheDocument();
  });

  it("Shows error-messages if fields are empty and form is submitted", async () => {
    fireEvent.click(screen.getByTestId("submitButton"));
    await waitFor(() => {
      expect(screen.getByText("Cart value is required")).toBeInTheDocument();
      expect(
        screen.getByText("Delivery distance is required")
      ).toBeInTheDocument();
      expect(
        screen.getByText("Amount of items is required")
      ).toBeInTheDocument();
      expect(screen.getByText("Valid time is required")).toBeInTheDocument();
    });
  });

  it("Should not submit formdata if a field value is invalid", async () => {
    fireEvent.change(screen.getByTestId("cartValue"), {
      target: { value: MOCK_INVALID_VALUES.cartValue },
    });
    fireEvent.change(screen.getByTestId("deliveryDistance"), {
      target: { value: MOCK_INVALID_VALUES.deliveryDistance },
    });
    fireEvent.change(screen.getByTestId("numberOfItems"), {
      target: { value: MOCK_INVALID_VALUES.numberOfItems },
    });
    fireEvent.change(screen.getByTestId("orderTime"), {
      target: { value: MOCK_INVALID_VALUES.orderTime },
    });

    fireEvent.click(screen.getByTestId("submitButton"));

    await waitFor(() => {
      expect(mockSaveCalculationResults).not.toHaveBeenCalled();
    });
  });

  it("Should submit formdata if a field values are valid", async () => {
    fireEvent.change(screen.getByTestId("cartValue"), {
      target: { value: MOCK_VALID_VALUES.cartValue },
    });
    fireEvent.change(screen.getByTestId("deliveryDistance"), {
      target: { value: MOCK_VALID_VALUES.deliveryDistance },
    });
    fireEvent.change(screen.getByTestId("numberOfItems"), {
      target: { value: MOCK_VALID_VALUES.numberOfItems },
    });
    fireEvent.change(screen.getByTestId("orderTime"), {
      target: { value: MOCK_VALID_VALUES.orderTime },
    });

    fireEvent.click(screen.getByTestId("submitButton"));

    await waitFor(() => {
      expect(mockSaveCalculationResults).toHaveBeenCalled();
    });
  });
});
