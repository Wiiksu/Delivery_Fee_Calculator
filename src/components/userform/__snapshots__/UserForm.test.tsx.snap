// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<UserForm /> > Should render correctly to the snapshot 1`] = `
<DocumentFragment>
  <section
    class="bg-purple-400 px-6 pb-6 pt-2 rounded shadow-lg h-fit w-[22rem] md:w-[27rem]"
  >
    <h2
      class="text-center text-purple-50 text-2xl font-bold py-2 mt-3 mb-4 bg-purple-500 border-2 border-purple-50 border-opacity-60 rounded-md select-none"
    >
      Order Details
    </h2>
    <form
      class="flex flex-col gap-1"
      novalidate=""
    >
      <section>
        <div
          class="relative flex w-full items-center"
        >
          <label
            class="text-purple-50 font-semibold w-[6rem] md:w-40 leading-5 md:leading-none text-base md:text-lg"
          >
            Cart Value:
          </label>
          <input
            class="w-52 rounded py-1 px-2 outline-none box-border border-collapse border-2 border-purple-50 border-opacity-60 focus:border-purple-500 focus:border-2"
            data-testid="cartValue"
            name="cartValue"
            type="text"
          />
        </div>
      </section>
      <section>
        <div
          class="relative flex w-full items-center"
        >
          <label
            class="text-purple-50 font-semibold w-[6rem] md:w-40 leading-5 md:leading-none text-base md:text-lg"
          >
            Delivery distance:
          </label>
          <input
            class="w-52 rounded py-1 px-2 outline-none box-border border-collapse border-2 border-purple-50 border-opacity-60 focus:border-purple-500 focus:border-2"
            data-testid="deliveryDistance"
            name="deliveryDistance"
            type="text"
          />
        </div>
      </section>
      <section>
        <div
          class="relative flex w-full items-center"
        >
          <label
            class="text-purple-50 font-semibold w-[6rem] md:w-40 leading-5 md:leading-none text-base md:text-lg"
          >
            Amount of items:
          </label>
          <input
            class="w-52 rounded py-1 px-2 outline-none box-border border-collapse border-2 border-purple-50 border-opacity-60 focus:border-purple-500 focus:border-2"
            data-testid="numberOfItems"
            name="numberOfItems"
            type="text"
          />
        </div>
      </section>
      <section>
        <div
          class="relative flex w-full items-center"
        >
          <label
            class="text-purple-50 font-semibold w-[6rem] md:w-40 leading-5 md:leading-none text-base md:text-lg"
          >
            Time:
          </label>
          <input
            class="w-52 rounded py-1 px-2 outline-none box-border border-collapse border-2 border-purple-50 border-opacity-60 focus:border-purple-500 focus:border-2"
            data-testid="orderTime"
            name="orderTime"
            type="datetime-local"
          />
        </div>
      </section>
      <div
        class="flex flex-col mt-2 gap-3"
      >
        <button
          class="bg-purple-500 text-purple-50 text-lg border-2 border-purple-50 border-opacity-60 font-semibold py-1 px-2 rounded-md hover:bg-purple-600"
          data-testid="submitButton"
          type="submit"
        >
          Calculate Delivery Fee
        </button>
      </div>
    </form>
  </section>
</DocumentFragment>
`;
